proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+
+Use cases:
+
+- Unsigned types
+```kotlin
+inline class UInt(private val value: Int) { ... }
+inline class UShort(private val value: Short) { ... }
+inline class UByte(private val value: Byte) { ... }
+inline class ULong(private val value: Long) { ... }
+```
+
+- Native types like `size_t` for Kotlin/Native
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
+    - String enum for Kotlin/JS (see [WebIDL enums](https://www.w3.org/TR/WebIDL-1/#idl-enums))
+- Units of measurement
+- Result type (aka Try monad) [KT-18608](https://youtrack.jetbrains.com/issue/KT-18608)
+- Inline property delegates
+```kotlin
+class A {
+    var something by InlinedDelegate(Foo()) // no actual instantiation of `InlinedDelegate`
+}
+
+
+inline class InlinedDelegate<T>(var node: T) {
+    operator fun setValue(thisRef: A, property: KProperty<*>, value: T) {
+        if (node !== value) {
+            thisRef.notify(node, value)
+        }
+        node = value
+    }
+
+    operator fun getValue(thisRef: A, property: KProperty<*>): T {
+        return node
+    }
+}
+```
+
+- Inline wrappers
+    - Typed wrappers
+    ```kotlin
+    inline class Name(private val s: String)
+    inline class Password(private val s: String)
+    
+    fun foo() {
+        var n = Name("n") // no actual instantiation, on JVM type of `n` is String
+        val p = Password("p")
+        n = "other" // type mismatch error
+        n = p // type mismatch error
+    }
+    ```
+
+    - API refinement
+    ```
+    // Java
+    public class Foo {
+        public Object[] objects() { ... }
+    }
+    
+    // Kotlin
+    inline class RefinedFoo(val f: Foo) {
+        inline fun <T> array(): Array<T> = f.objects() as Array<T>
+    }
+    ``` 
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```kotlin
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```kotlin
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
```

Comment:
What is the reason for this restriction?

Reply:
@ricmf It can only work on a single reference. If you have more than one property, you need two references, and that only works with a real enclosing instance (that is allocated, not inline). However, you can perfectly make an inline class enclosing a `Pair` if it makes sense.

Reply:
Maybe it is possible to use multiple values, I am thinking about it.
It should work like JPA `Embeddable` class , JVM escape analysis should avoid extra allocation, probably this is out of scope and we should wait for Valhalla.

The problem become tricky in presence of annotations.